pyccel/codegen/printing/cwrappercode.py

```diff
@@ -0,0 +1,351 @@
+
+from pyccel.ast.datatypes import NativeInteger, NativeBool, NativeComplex, NativeReal
+
+from pyccel.ast.cwrapper import PyccelPyObject, PyArg_ParseTupleNode, PyBuildValueNode
+from pyccel.ast.cwrapper import PyArgKeywords, FuncCall
+from pyccel.ast.cwrapper import Py_True, Py_False
+
+from pyccel.ast.type_inference import str_dtype
+
+from pyccel.errors.errors import Errors
+from pyccel.errors.messages import *
+
+errors = Errors()
+
+__all__ = ["CWrapperCodePrinter", "cwrappercode"]
+
+dtype_registry = {('pyobject', 0) : 'PyObject'}
+
+class CWrapperCodePrinter(CCodePrinter):
+    def __init__(self, parser, settings={}):
+        CCodePrinter.__init__(self, parser,settings)
+        self._cast_functions_dict = OrderedDict()
+        self._function_wrapper_names = dict()
+        self._global_names = set()
+
+    def get_new_name(self, used_names, requested_name):
+        if requested_name not in used_names:
+            used_names.add(requested_name)
+            return requested_name
+        else:
+            incremented_name, _ = create_incremented_string(used_names, prefix=requested_name)
+            return incremented_name
+
+    def find_in_dtype_registry(self, dtype, prec):
+        try :
+            return dtype_registry[(dtype, prec)]
+        except KeyError:
+            return CCodePrinter.find_in_dtype_registry(self, dtype, prec)
+
+    def get_cast_function(self, used_names, cast_type, from_variable, to_variable):
+        if cast_type in self._cast_functions_dict:
+            return self._cast_functions_dict[cast_type]
+
+        cast_function_name = self.get_new_name(used_names, cast_type)
+        cast_function_arg = [from_variable]
+        cast_function_result = [to_variable]
+        cast_function_local_vars = []
+
+        #switch case of cast_type
+        if cast_type == 'pyint_to_bool':
+            cast_function_body = [Assign(cast_function_result[0], Bool(cast_function_arg[0]))]
+
+        elif cast_type == 'bool_to_pyobj':
+            cast_function_body = [IfTernaryOperator((Bool(cast_function_arg[0]),
+                [AliasAssign(cast_function_result[0], Py_True)]),
+                (BooleanTrue(), [AliasAssign(cast_function_result[0], Py_False)]))]
+
+        elif cast_type == 'pycomplex_to_complex':
+            real_part = Variable(dtype = NativeReal(),
+                            name = self.get_new_name(used_names, 'real_part'))
+            imag_part = Variable(dtype = NativeReal(),
+                            name = self.get_new_name(used_names, 'imag_part'))
+            cast_function_local_vars += [real_part, imag_part]
+            cast_function_body = [Assign(real_part, FuncCall('PyComplex_RealAsDouble', cast_function_arg))]
+            cast_function_body += [Assign(imag_part, FuncCall('PyComplex_ImagAsDouble', cast_function_arg))]
+            cast_function_body += [Assign(cast_function_result[0], FuncCall('__builtin_complex', [real_part, imag_part]))]
```

Comment:
The function `__builtin_complex` is gcc specific. It does not necessarily exist on other compilers. Complex numbers should therefore be initialised in a different way. E.g. with maths and by defining `_print_ImaginaryUnit`
See : https://stackoverflow.com/a/6418824/5957195